BJS2-77103 Хранение запросов и операций

VladimirEkgardt opened the pull request.

VladimirEkgardt committed:
BJS2-77103 Хранение запросов и операций

KolyanGrom commented on src/main/java/faang/school/accountservice/controller/RequestController.java:
```diff
@@ -0,0 +1,24 @@
+
+    @PostMapping
+    public ResponseRequestDto createRequest(@Valid @RequestBody CreateRequestDto createRequestDto){
+        return requestService.createRequest(createRequestDto);
```
Сейчас метод возвращает 200 OK, но для создания ресурса правильнее использовать код ответа 201 Created

может переписать так?

@PostMapping
public ResponseEntity<ResponseRequestDto> createRequest(@Valid @RequestBody CreateRequestDto dto) {
ResponseRequestDto response = requestService.createRequest(dto);
return ResponseEntity.status(HttpStatus.CREATED).body(response);
}

KolyanGrom commented on src/main/java/faang/school/accountservice/enums/RequestStatus.java:
```diff
@@ -0,0 +1,9 @@
+    PROCESSED,
+    COMPLETED,
+    CANCELLED
+}
```
Может заменить на более читаемые и логически понятные статусы?

например READY_FOR_EXECUTION заменить на IN_PROGRESS (что будет обозначать "В процессе выполнения")

COMPLETED заменить на SUCCESS (Успешно завершен)

и добавить еще статус FAILED - Завеошен с ошибкой потому что CANCELLED озночяет что явно отменен

KolyanGrom commented on src/main/java/faang/school/accountservice/kafka/KafkaConsumerConfig.java:
```diff
@@ -0,0 +1,57 @@
+        config.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
+        config.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, JsonDeserializer.class);
+        config.put(JsonDeserializer.TRUSTED_PACKAGES, "*");
+        config.put(JsonDeserializer.VALUE_DEFAULT_TYPE, "faang.school.accountservice.dto.CreateRequestDto");
```
Давай я как ты мне тогда - вынесем это в конфиг?)))

trusted-packages: "faang.school.accountservice.dto"

и будем использовать так
config.put(JsonDeserializer.TRUSTED_PACKAGES, trustedPackages);

объявив в

@Value("${kafka.trusted-packages}")
private String trustedPackages;

KolyanGrom commented on src/main/java/faang/school/accountservice/listener/EventListener.java:
```diff
@@ -0,0 +1,19 @@
+    private final RequestService requestService;
+
+    @KafkaListener(topics = "${kafka.topics.cons-request}", groupId = "${kafka.group}")
+    public void listener(CreateRequestDto createRequestDto){
```
Добавить аннотацию @Valid чтобы проверить валидность DTO

KolyanGrom commented on src/main/java/faang/school/accountservice/listener/OpenCloseRequest.java:
```diff
@@ -0,0 +1,48 @@
+    @Async
+    @Scheduled(cron = "${cron.expression}")
+    public void openRequests() {
+        List<Request> requests = requestRepository.findAll();
```
findAll() загружает все записи из БД, даже если нужны только запросы с определенным статусом. Заменить на requestRepository.findByStatus(status).

KolyanGrom commented on src/main/java/faang/school/accountservice/listener/OpenCloseRequest.java:
```diff
@@ -0,0 +1,48 @@
+    @Async
+    @Scheduled(cron = "${cron.expression}")
+    public void closeRequests() {
+        List<Request> requests = requestRepository.findAll();
```
аналогично про поиск по всем запросам выше.

KolyanGrom commented on src/main/java/faang/school/accountservice/listener/OpenCloseRequest.java:
```diff
@@ -0,0 +1,48 @@
+            }
+        }
+    }
+}
```
общий вывод:

Оба метода почти идентичны. Нет случая для ошбки, если openRequest() или closeRequest() выбросят исключение, задача прервется без лога.

попробуй такую реализацию:

@Async
@Scheduled(cron = "${cron.open-expression}")
public void processReadyRequests() {
List<Request> requests = requestRepository.findByStatus(RequestStatus.READY_FOR_EXECUTION);
requests.forEach(request -> {
try {
requestService.openRequest(request);
log.info("Request opened: {}", request.getId());
} catch (Exception e) {
log.error("Failed to open request {}: {}", request.getId(), e.getMessage());
}
});
}

VladimirEkgardt committed:
1) RequestController: добавил responsEntity, со статусом 201
2) RequestStatus: READY_FOR_EXECUTION->IN_PROGRESS, COMPLETED->SUCCESS, добавил FAILED
3) Вынес в пропертис trustedPackages
4) EventListener на поле createRequestDto повесил @Valid
5) RequestRepository добавил findAllByStatus
6) OpenCloseRequest переписал методы openRequests, closeRequests с новым методом из репозитория. Так же добавил обработку ошибок

SnurnicynKonstantin commented on src/main/java/faang/school/accountservice/kafka/KafkaConsumerConfig.java:
```diff
@@ -0,0 +1,60 @@
+@Configuration
+@RequiredArgsConstructor
+@EnableKafka
+public class KafkaConsumerConfig {
```
Конфиги должны находиться в пакете config

SnurnicynKonstantin commented on src/main/java/faang/school/accountservice/repository/RequestRepository.java:
```diff
@@ -0,0 +1,27 @@
+public interface RequestRepository extends JpaRepository<Request, Long> {
+
+    @Query(nativeQuery = true, value = """
+            SELECT * FROM requests WHERE idempotent_token = ?1 FOR UPDATE
```
Блокировку использовал, интересно

SnurnicynKonstantin commented on src/main/java/faang/school/accountservice/enums/RequestType.java:
```diff
@@ -0,0 +1,6 @@
+package faang.school.accountservice.enums;
+
+public enum RequestType {
```
Запрос и операция это одно и то же. 
Здесь надо абстрактных операций добавить. Например SEND_NOTIFICATION, PRINT_DOCUMENT, любы

VladimirEkgardt committed:
1) Переместил KafkaConsumerConfig в пакет config
2) Добавил операции в RequestType
